sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala

```diff
@@ -1140,9 +1167,26 @@ private class FullOuterIterator(
 
   override def advanceNext(): Boolean = {
     val r = smjScanner.advanceNext()
-    if (r) numRows += 1
+    if (r) {
+      numRows += 1
+    } else {
+      smjScanner.close()
+    }
     r
   }
 
   override def getRow: InternalRow = resultProj(joinedRow)
 }
+
+trait CloseableScanner {
+  def closeIterator(iter: RowIterator): Unit = {
+    if (iter.isInstanceOf[RowIteratorFromScala]) {
+      val rowIter = iter.asInstanceOf[RowIteratorFromScala]
+      val underlyingIter = rowIter.toScala
```

Comment:
Does `.toScala` return the underlying thing or make a wrapper? Just checking, haven't investigated it, but want to make sure this closes what we think.

Reply:
For our case of RowIteratorFromScala, this would return the underlying iterator